### PR TITLE
[NCL-7220] Implement Change Quality action

### PR DESCRIPTION
**Pull request by patrikk0123**

### CODE REVIEW

File: src/components/ActionModal/ActionModal.tsx [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+  actionTitle,
+  isOpen,
+  isConfirmActionDisabled,
+  onToggle,
+  onSubmit,
+  serviceContainerCreatingUpdating,
+  modalVariant,
+  onSuccessActions,
+}: PropsWithChildren<IActionModalProps>) => {
+  const refresh = useRefresh();
```
[the remark this answers is not in the document]
Comment (patrikk0123) [reply src/components/ActionModal/ActionModal.tsx:51]: I will stick to the hook name.

File: src/components/ActionModal/ActionModal.tsx [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+export const ActionModal = ({
+  children,
+  modalTitle,
+  actionTitle,
+  isOpen,
+  isConfirmActionDisabled,
+  onToggle,
+  onSubmit,
+  serviceContainerCreatingUpdating,
+  modalVariant,
```
[the remark this answers is not in the document]
Comment (patrikk0123) [reply src/components/ActionModal/ActionModal.tsx:48]: The only option to apply the default value is the `modalVariant` parameter, but I don't see any value that would be good as 'default'. For example, on 'Mark Milestone as current' modal, this will be `small`, while on 'Artifact edit quality', it is `large`.

File: src/components/ActionModal/ActionModal.tsx [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+  const onClose = () => {
+    if (wasActionSuccessful) {
+      refresh();
+    } else {
+      onToggle();
+    }
+  };
+
+  return (
+    <Modal
```
[the remark this answers is not in the document]
Comment (patrikk0123) [reply src/components/ActionModal/ActionModal.tsx:65]: The parameters used here seem to be fine also in PatternFly v5.

File: src/components/ActionModal/ActionModal.tsx [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+        <Button
+          variant="primary"
+          onClick={onSubmit}
+          isDisabled={isConfirmActionDisabled || serviceContainerCreatingUpdating.loading}
+        >
+          {!serviceContainerCreatingUpdating.error && (
+            <ServiceContainerLoading variant="icon" {...serviceContainerCreatingUpdating} title={actionTitle} />
+          )}{' '}
+          {actionTitle}
+        </Button>,
```
[the remark this answers is not in the document]
Comment (patrikk0123) [reply src/components/ActionModal/ActionModal.tsx:79]: Your idea is applicable:
- `onSubmit` and `isConfirmActionDisabled` would be removed
- `actionTitle` is used as the title of `ServiceContainerCreatingUpdating` but that is not that much of a problem;  instead some generic text or `modalTitle` could replace the tile of Service Container
- `onSuccessActions` would be replaced by `actions` to which all the action buttons would be passed - including the progress button (but the `wasActionSuccessful` variable logic would need to go to the component one level up)

File: src/components/ArtifactEditQualityModal/ArtifactEditQualityModal.tsx [hunk 5]
```diff
@@ -0,0 +1,181 @@
...
+};
+
+interface IArtifactEditQualityModalProps {
+  artifact: Artifact;
+  isModalOpen: boolean;
+  toggleModal: () => void;
+  variant: 'page' | 'list';
+}
+
+export const ArtifactEditQualityModal = ({ artifact, isModalOpen, toggleModal, variant }: IArtifactEditQualityModalProps) => {
```
[the remark this answers is not in the document]
Comment (matedo1) [reply src/components/ArtifactEditQualityModal/ArtifactEditQualityModal.tsx]: done

File: src/components/ActionModal/ActionModal.tsx (hunk 2 above)
[the remark this answers is not in the document]
Comment (matedo1) [reply src/components/ActionModal/ActionModal.tsx:48]: done

File: src/hooks/useRefresh.ts [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+ */
+export const useRefresh = () => {
+  const timer: MutableRefObject<NodeJS.Timer | undefined> = useRef();
+
+  const navigate = useNavigate();
+  const location = useLocation();
+
+  const refresh = () => {
+    const originalRoute = location.pathname + location.search;
+    console.log(originalRoute);
```
[the remark this answers is not in the document]
Comment (matedo1) [reply src/hooks/useRefresh.ts]: done

File: src/components/ActionModal/ActionModal.tsx [hunk 7]
```diff
@@ -0,0 +1,105 @@
...
+ * @param serviceContainerCreatingUpdating - service container of the confirm action
+ * @param modalVariant - modal size variant
+ * @param onSuccessActions - additional action buttons to be displayed after susccessful confirm action
+ */
+export const ActionModal = ({
+  children,
+  modalTitle,
+  actionTitle,
+  isOpen,
+  isConfirmActionDisabled,
```
[the remark this answers is not in the document]
Comment (matedo1) [reply src/components/ActionModal/ActionModal.tsx]: done

File: src/components/ArtifactEditQualityModal/ArtifactEditQualityModal.tsx [hunk 8]
```diff
@@ -0,0 +1,181 @@
...
+      isOpen={isModalOpen}
+      isConfirmActionDisabled={isSubmitDisabled}
+      onToggle={toggleModal}
+      onSubmit={onSubmit}
+      serviceContainerCreatingUpdating={serviceContainerArtifactEditQuality}
+      modalVariant="large"
+      onSuccessActions={
+        variant === 'list'
+          ? [
+              <Button
```
[the remark this answers is not in the document]
Comment (matedo1) [reply src/components/ArtifactEditQualityModal/ArtifactEditQualityModal.tsx]: done

File: src/hooks/useRefresh.ts [hunk 9]
```diff
@@ -0,0 +1,30 @@
...
+
+const EMPTY_ROUTE = '/';
+
+/**
+ * Hook returning 'refresh' function.
+ * Refresh function refreshes the current route using the SPA approach.
+ * Original query parameters are remembered. History is NOT altered.
+ *
+ * inspired by: https://dev.to/zbmarius/react-route-refresh-without-page-reload-1907
+ */
```
[the remark this answers is not in the document]
Comment (matedo1) [reply src/hooks/useRefresh.ts:8]: done

File: src/hooks/useRefresh.ts [hunk 10]
```diff
@@ -0,0 +1,30 @@
...
+export const useRefresh = () => {
+  const timer: MutableRefObject<NodeJS.Timer | undefined> = useRef();
```
[the remark this answers is not in the document]
Comment (matedo1) [reply src/hooks/useRefresh.ts]: done

File: src/hooks/useRefresh.ts [hunk 11]
```diff
@@ -0,0 +1,30 @@
...
+  const location = useLocation();
+
+  const refresh = () => {
+    const originalRoute = location.pathname + location.search;
+    console.log(originalRoute);
+
+    navigate(EMPTY_ROUTE, { replace: true });
+    timer.current = setTimeout(() => {
+      navigate(originalRoute, { replace: true });
+    }, 10);
```
[the remark this answers is not in the document]
Comment (matedo1) [reply src/hooks/useRefresh.ts]: done

File: src/components/ActionModal/ActionModal.tsx (hunk 1 above)
[the remark this answers is not in the document]
Comment (matedo1) [reply src/components/ActionModal/ActionModal.tsx:51]: ok

File: src/components/ActionModal/ActionModal.tsx [hunk 12]
```diff
@@ -0,0 +1,101 @@
...
+      actions={[
+        <Button variant="primary" onClick={onSubmit} isDisabled={isSubmitDisabled || serviceContainerCreatingUpdating.loading}>
+          {!serviceContainerCreatingUpdating.error && (
+            <ServiceContainerLoading variant="icon" {...serviceContainerCreatingUpdating} title={actionTitle} />
+          )}{' '}
+          {actionTitle}
+        </Button>,
+        ...(wasActionSuccessful && onSuccessActions ? onSuccessActions : []),
+        <Button variant="link" onClick={onClose}>
+          {wasActionSuccessful ? 'Close and refresh' : 'Cancel'}
```
Comment (matedo1) [inline src/components/ActionModal/ActionModal.tsx:78]: There is one more scenario, when this is not working properly:
- first request is successful (ie. NEW -> DELETED)
- second request fails (ie. DELETED -> NEW  // invalid operation)

Then user can see only "Cancel" and Close icon, in both cases, the user will be returned to the deprecated page.

Solution would be to keep one more useState variable that would be marked as `true` when `wasActionSuccessful` is true for the first time.

So basically there will be two states captured:

- `wasLastActionSuccessful` (use instead of `wasActionSuccessful`)
- `wasAnyActionSuccessful`

**Review by matedo1 — CHANGES_REQUESTED** [review]

File: src/components/ActionModal/ActionModal.tsx [hunk 13]
```diff
@@ -0,0 +1,101 @@
...
+      showClose={false}
+    >
+      <div className="m-r-0">
+        <ServiceContainerCreatingUpdating {...serviceContainerCreatingUpdating} title={actionTitle}>
+          <div className={css('p-t-global', 'p-l-global', 'p-r-global')}>
+            <div className={css('p-b-global', styles['action-modal-header'])}>
+              <TextContent className={styles['action-modal-title']}>
+                <Text component={TextVariants.h1}>{modalTitle}</Text>
+              </TextContent>
+              <Button variant="plain" onClick={onClose} className={styles['action-modal-close-button']}>
```
[the remark this answers is not in the document]
Comment (patrikk0123) [reply src/components/ActionModal/ActionModal.tsx:91]: I used `margin-top: 15px`.

File: src/components/ArtifactEditQualityModalButton/ArtifactEditQualityModalButton.tsx [hunk 14]
```diff
@@ -0,0 +1,174 @@
...
+  qualityLevelReason: {
+    isRequired: true,
+  },
+};
+
+interface IArtifactEditQualityModalProps {
+  artifact: Artifact;
+  isModalOpen: boolean;
+  toggleModal: () => void;
+  variant: 'page' | 'list';
```
[the remark this answers is not in the document]
Comment (patrikk0123) [reply src/components/ArtifactEditQualityModalButton/ArtifactEditQualityModalButton.tsx:57]: Right, but I think it should be the other way around.

File: src/components/ArtifactEditQualityModalButton/ArtifactEditQualityModalButton.tsx [hunk 15]
```diff
@@ -0,0 +1,174 @@
...
+import { useServiceContainer } from 'hooks/useServiceContainer';
+
+import { ActionModal } from 'components/ActionModal/ActionModal';
+import { ProtectedComponent } from 'components/ProtectedContent/ProtectedComponent';
+
+import * as artifactApi from 'services/artifactApi';
+
+interface IArtifactEditQualityModalButtonProps {
+  artifact: Artifact;
+  variant: 'page' | 'list';
```
[the remark this answers is not in the document]
Comment (patrikk0123) [reply src/components/ArtifactEditQualityModalButton/ArtifactEditQualityModalButton.tsx:20]: `'secondary' | 'tertiary'` is confusing

File: src/components/ActionModal/ActionModal.tsx [hunk 16]
```diff
@@ -0,0 +1,101 @@
...
+    }
+  };
+
+  return (
+    <Modal
+      variant={modalVariant}
+      isOpen={isOpen}
+      onClose={onClose}
+      actions={[
+        <Button variant="primary" onClick={onSubmit} isDisabled={isSubmitDisabled || serviceContainerCreatingUpdating.loading}>
```
[the remark this answers is not in the document]
Comment (patrikk0123) [reply src/components/ActionModal/ActionModal.tsx:70]: it disabled during the loading state, it's obvious why so.

isSubmitDisabled is true when:
- form was not touched (which is intuitive)
- form has validation errors (messages for which are displayed on the respective field level)
- required field was not filled (there are asterisks for these inputs)

File: src/components/ActionModal/ActionModal.tsx [hunk 17]
```diff
@@ -0,0 +1,101 @@
...
+            <ServiceContainerLoading variant="icon" {...serviceContainerCreatingUpdating} title={actionTitle} />
+          )}{' '}
+          {actionTitle}
+        </Button>,
+        ...(wasActionSuccessful && onSuccessActions ? onSuccessActions : []),
+        <Button variant="link" onClick={onClose}>
+          {wasActionSuccessful ? 'Close and refresh' : 'Cancel'}
+        </Button>,
+      ]}
+      hasNoBodyWrapper
```
[the remark this answers is not in the document]
Comment (matedo1) [reply src/components/ActionModal/ActionModal.tsx:92]: done

File: src/components/ActionModal/ActionModal.tsx [hunk 18]
```diff
@@ -0,0 +1,101 @@
...
+ *
+ * @param modalTitle - title of the modal
+ * @param actionTitle - title of the confirm action button
+ * @param isOpen - is the modal open?
+ * @param isSubmitDisabled - is the confirm action button disabled?
+ * @param onToggle - function toggling the modal visibility
+ * @param onSubmit - confirm action button callback
+ * @param serviceContainerCreatingUpdating - service container of the confirm action
+ * @param modalVariant - modal size variant
+ * @param onSuccessActions - additional action buttons to be displayed after susccessful confirm action
```
[the remark this answers is not in the document]
Comment (matedo1) [reply src/components/ActionModal/ActionModal.tsx]: done

File: src/components/ActionModal/ActionModal.tsx [hunk 19]
```diff
@@ -0,0 +1,101 @@
...
+    <Modal
+      variant={modalVariant}
+      isOpen={isOpen}
+      onClose={onClose}
+      actions={[
+        <Button variant="primary" onClick={onSubmit} isDisabled={isSubmitDisabled || serviceContainerCreatingUpdating.loading}>
+          {!serviceContainerCreatingUpdating.error && (
+            <ServiceContainerLoading variant="icon" {...serviceContainerCreatingUpdating} title={actionTitle} />
+          )}{' '}
+          {actionTitle}
```
[the remark this answers is not in the document]
Comment (matedo1) [reply src/components/ActionModal/ActionModal.tsx]: done

File: src/components/ActionModal/ActionModal.tsx [hunk 20]
```diff
@@ -0,0 +1,101 @@
...
+        </Button>,
+        ...(wasActionSuccessful && onSuccessActions ? onSuccessActions : []),
+        <Button variant="link" onClick={onClose}>
+          {wasActionSuccessful ? 'Close and refresh' : 'Cancel'}
+        </Button>,
+      ]}
+      hasNoBodyWrapper
+      showClose={false}
```
[the remark this answers is not in the document]
Comment (matedo1) [reply src/components/ActionModal/ActionModal.tsx:94]: done

File: src/components/ArtifactEditQualityModalButton/ArtifactEditQualityModalButton.tsx (hunk 15 above)
[the remark this answers is not in the document]
Comment (matedo1) [reply src/components/ArtifactEditQualityModalButton/ArtifactEditQualityModalButton.tsx]: done

File: src/components/ArtifactEditQualityModalButton/ArtifactEditQualityModalButton.tsx (hunk 14 above)
[the remark this answers is not in the document]
Comment (matedo1) [reply src/components/ArtifactEditQualityModalButton/ArtifactEditQualityModalButton.tsx]: done

File: src/components/ActionModal/ActionModal.tsx [hunk 21]
```diff
@@ -0,0 +1,101 @@
...
+import styles from './ActionModal.module.css';
+
+interface IActionModalProps {
+  modalTitle: string;
+  actionTitle: string;
+  isOpen: boolean;
+  isSubmitDisabled: boolean;
+  onToggle: () => void;
+  onSubmit: () => void;
+  serviceContainerCreatingUpdating: IServiceContainer;
```
[the remark this answers is not in the document]
Comment (matedo1) [reply src/components/ActionModal/ActionModal.tsx]: done

File: src/components/ActionModal/ActionModal.tsx (hunk 16 above)
[the remark this answers is not in the document]
Comment (matedo1) [reply src/components/ActionModal/ActionModal.tsx]: ok

File: src/hooks/useServiceContainer.ts [hunk 22]
```diff
@@ -45,7 +45,7 @@ export const DataValues = {
  * @param service - Service to be executed to load data
  * @returns Object with data, loading and error property
  */
-export const useServiceContainer = (service: Function): IServiceContainer => {
+export const useServiceContainer = (service: Function, loadingStateDelay: number = 200): IServiceContainer => {
```
Comment (matedo1) [inline src/hooks/useServiceContainer.ts:48]: Can you please add also TSDoc?

**Review by matedo1 — CHANGES_REQUESTED** [review]

File: src/components/ActionModal/ActionModal.tsx (hunk 4 above)
[the remark this answers is not in the document]
Comment (matedo1) [reply src/components/ActionModal/ActionModal.tsx]: We agreed to implement progressButton later, resolving this

File: src/components/ActionModal/ActionModal.tsx (hunk 13 above)
[the remark this answers is not in the document]
Comment (matedo1) [reply src/components/ActionModal/ActionModal.tsx:103]: done

File: src/hooks/useServiceContainer.ts (hunk 22 above)
Comment (matedo1) [reply src/hooks/useServiceContainer.ts:49]: done

**Review by matedo1 — CHANGES_REQUESTED** [review]
I noticed one more issue during the local testing.

Steps:
- when previous http request will result in error
- next update is missing loading spinner inside button

**Review by matedo1 — APPROVED** [review]
